test/unit/pipeline/FetchUnitTest.cc

```diff
@@ -442,5 +462,288 @@ TEST_F(PipelineFetchUnitTest, idleLoopBufferDueToNotTakenBoundary) {
+  // so the test is only relevent in the case where they are different
+  if (insnMinSizeBytes < insnMaxSizeBytes) {
+    ON_CALL(isa, getMaxInstructionSize())
+        .WillByDefault(Return(insnMaxSizeBytes));
+    ON_CALL(isa, getMinInstructionSize())
+        .WillByDefault(Return(insnMinSizeBytes));
+    ON_CALL(memory, getCompletedReads()).WillByDefault(Return(completedReads));
+
+    // Buffer will contain invalid min bytes so predecode returns 0 bytes read
+    ON_CALL(isa, predecode(_, insnMinSizeBytes, 0x10 - insnMinSizeBytes, _))
+        .WillByDefault(Return(0));
+
+    // getMaxInstructionSize called for second time in assertion
+    if (strcmp(SIMENG_BUILD_TYPE, "Release") == 0) {
+      EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    } else {
+      EXPECT_CALL(isa, getMaxInstructionSize()).Times(2);
+    }
+    EXPECT_CALL(memory, getCompletedReads()).Times(1);
+    EXPECT_CALL(isa, getMinInstructionSize()).Times(1);
+    EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+    uint64_t setPC = blockSize - insnMinSizeBytes;
+    // Fetch a minimum number of bytes, buffered bytes = 0
+    fetchUnit.updatePC(setPC);
+    // Tick
+    fetchUnit.tick();
+
+    // No data consumed
+    EXPECT_EQ(fetchUnit.bufferedBytes_, insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.pc_, blockSize - insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], MacroOp());
+
+    // Expect that memory is requested even though there is data in the buffer
+    // as bufferedBytes < maxInstructionSize
+    EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    EXPECT_CALL(memory,
+                requestRead(Field(&memory::MemoryAccessTarget::address, 16), _))
+        .Times(1);
+    fetchUnit.requestFromPC();
+
+    EXPECT_EQ(fetchUnit.bufferedBytes_, insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.pc_, blockSize - insnMinSizeBytes);
+
+    // Memory doesn't complete reads in next cycle but buffered bytes should
+    // attempt to be predecoded
+    ON_CALL(memory, getCompletedReads())
+        .WillByDefault(Return(span<memory::MemoryReadResult>{nullptr, 0}));
+    // Predecode still returns no bytes read
+    ON_CALL(isa, predecode(_, insnMinSizeBytes, 0x10 - insnMinSizeBytes, _))
+        .WillByDefault(Return(0));
+
+    // getMaxInsnSize called again in assertion
+    if (strcmp(SIMENG_BUILD_TYPE, "Release") == 0) {
+      EXPECT_CALL(isa, getMaxInstructionSize()).Times(1);
+    } else {
+      EXPECT_CALL(isa, getMaxInstructionSize()).Times(2);
+    }
+    EXPECT_CALL(memory, getCompletedReads()).Times(1);
+    EXPECT_CALL(isa, getMinInstructionSize()).Times(2);
+    EXPECT_CALL(isa, predecode(_, _, _, _)).Times(1);
+
+    // Tick
+    fetchUnit.tick();
+
+    // Ensure min bytes are not consumed
+    EXPECT_EQ(fetchUnit.bufferedBytes_, insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.pc_, blockSize - insnMinSizeBytes);
+    EXPECT_EQ(fetchUnit.output_.getTailSlots()[0], MacroOp());
```

Comment:
I think rather than address 16, we should base the next fetch address off of the block size